overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,232 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi"
+import { Bucket } from "./bucket";
+import { BucketNotification } from "./bucketNotification";
+import * as lambda from "../lambda"
+
+/**
+ * Arguments to help customize a notification subscription for a bucket.
+ */
+export interface CommonBucketSubscriptionArgs {
+    /**
+     * An optional prefix or suffix to filter down notifications.  See
+     * aws.s3.BucketNotification.lambdaFunctions for more details.
+     */
+    filterPrefix?: string;
+    filterSuffix?: string;
+}
+
+export interface BucketEventSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    /**
+     * Events to subscribe to. For example: "[s3:ObjectCreated:*]".  Cannot be empty.
+     */
+    events: string[];
+}
+
+/**
+ * Arguments to specifically control a subscription to 'ObjectCreated' notifications on a bucket.If
+ * more events than just 'ObjectCreated' events are desired, the 'subscribe' function should be used
+ * instead.
+ */
+export interface ObjectCreatedSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    // The type of event to listen for.  '*' is the default if not provided.
+    event?: "*" | "Put" | "Post" | "Copy" | "CompleteMultipartUpload";
```

Comment:
Might be nice to document each option e.g. from https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html.

Reply:
In fact, any reason not to have an enum of options here? Seems nicer to be able to say

`lambda.ObjectCreatedEvent` versus `"*"`

and similarly `lambda.ObjectCreatedPutEvent` versus `"Put"`, especially given case sensitivity.

Reply:
> and similarly lambda.ObjectCreatedPutEvent versus "Put", especially given case sensitivity.

Note: case sensitivity isn't really a problem.  TS will complain if you give hte wrong casing as this code explicitly defines the allowed string values you can provide.  

We can use an enum here though. I don't feeel strongly about that one way or the other.  Let me go look at how that would look.

Reply:
TypeScript will also give you intellisense here, allowing you to complete out to these names easily.  So it's not like having an enum actually makes things easier to type, or easier to check.

Reply:
This form is actually how TS does things with their library files, and is idiomatic for the TS community.  i.e.:  `    normalize(form: "NFC" | "NFD" | "NFKC" | "NFKD"): string;`  There's no need for an enum indirection as the "union" and "string" (or number) types subsume that.